### PR TITLE
Add keep option to distinct nvbench

### PR DESCRIPTION
**Pull request by bdice**
## Description
This PR adopts some work from @srinivasyadav18 with additional modifications. This is meant to complement #16484.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/benchmarks/stream_compaction/distinct.cpp [hunk 1]
```diff
@@ -40,27 +69,29 @@ void nvbench_distinct(nvbench::state& state, nvbench::type_list<Type>)
...
-using data_type = nvbench::type_list<bool, int8_t, int32_t, int64_t, float, cudf::timestamp_ms>;
+using data_type = nvbench::type_list<int32_t, int64_t>;
 
 NVBENCH_BENCH_TYPES(nvbench_distinct, NVBENCH_TYPE_AXES(data_type))
   .set_name("distinct")
   .set_type_axes_names({"Type"})
-  .add_int64_axis("NumRows", {10'000, 100'000, 1'000'000, 10'000'000});
+  .add_string_axis("keep", {"any", "first", "last", "none"})
+  .add_int64_axis("cardinality",
+                  {100, 1'000, 10'000, 100'000, 1'000'000, 10'000'000, 100'000'000, 1'000'000'000})
```
Comment (srinivasyadav18) [inline cpp/benchmarks/stream_compaction/distinct.cpp:84]: We can just decrease the default values in the axis.

File: cpp/benchmarks/stream_compaction/distinct.cpp [hunk 2]
```diff
@@ -40,27 +69,29 @@ void nvbench_distinct(nvbench::state& state, nvbench::type_list<Type>)
...
+  .add_int64_axis("NumRows",
+                  {100, 1'000, 10'000, 100'000, 1'000'000, 10'000'000, 100'000'000, 1'000'000'000});
```
Comment (PointKernel) [inline cpp/benchmarks/stream_compaction/distinct.cpp:86]: we adding many more benchmark cases by doing so. maybe worth considering fewer cases for a shorter default execution time
```suggestion
  .add_string_axis("keep", {"any", "first", "last", "none"})
  .add_int64_axis("cardinality",
                  {100, 100'000, 10'000'000, 1'000'000'000})
  .add_int64_axis("NumRows",
                  {100, 100'000, 10'000'000, 1'000'000'000});
```
not a hard requirement though

**Review by PointKernel — APPROVED** [review]
One non-blocking suggestion

LGTM.

File: cpp/benchmarks/stream_compaction/distinct.cpp [hunk 3]
```diff
@@ -23,15 +24,29 @@
...
 {
-  cudf::size_type const num_rows = state.get_int64("NumRows");
+  cudf::size_type const num_rows    = state.get_int64("NumRows");
+  auto const keep                   = get_keep(state.get_string("keep"));
+  cudf::size_type const cardinality = state.get_int64("cardinality");
+
+  if (cardinality > num_rows) {
+    state.skip("cardinality > num_rows");
+    return;
+  }
```
Comment (GregoryKimball) [inline cpp/benchmarks/stream_compaction/distinct.cpp:41]: I would prefer to omit this skipping condition. I recognize that we can't have 1M distinct elements in 1K rows, but this condition adds a lot of friction when sweeping NumRows for the high cardinality case. It forces me to run the full factorial and filter the outputs.

Comment (bdice) [reply cpp/benchmarks/stream_compaction/distinct.cpp:41]: Hmm. @GregoryKimball I reviewed the NVBench docs and I don't see a way to filter out certain jobs except by skipping them. https://github.com/NVIDIA/nvbench/blob/main/docs/benchmarks.md#beware-combinatorial-explosion-is-lurking

We might be able to use a string axis like `{"100,100", "100,1000", ..., "1000000000,1000000000"}` but that's hard to maintain.

Comment (PointKernel) [reply cpp/benchmarks/stream_compaction/distinct.cpp:41]: >  I don't see a way to filter out certain jobs except by skipping them.

https://github.com/NVIDIA/nvbench/pull/80 will solve this issue.

File: cpp/benchmarks/stream_compaction/distinct.cpp (hunk 2 above)
Comment (GregoryKimball) [reply cpp/benchmarks/stream_compaction/distinct.cpp]: @bdice I support @PointKernel 's suggestion. `distinct` just took me >15 minutes to run due to the wider axes.

**Comment by GregoryKimball** [comment]
Thanks guys, here is a performance snapshot of these benchmarks on A100
image

I would like to merge this update ASAP and then have @srinivasyadav18 pull the changes into https://github.com/rapidsai/cudf/pull/16484

File: cpp/benchmarks/stream_compaction/stable_distinct.cpp [hunk 4]
```diff
@@ -40,27 +55,29 @@ void nvbench_stable_distinct(nvbench::state& state, nvbench::type_list<Type>)
...
 
 NVBENCH_BENCH_TYPES(nvbench_stable_distinct, NVBENCH_TYPE_AXES(data_type))
   .set_name("stable_distinct")
   .set_type_axes_names({"Type"})
-  .add_int64_axis("NumRows", {10'000, 100'000, 1'000'000, 10'000'000});
+  .add_string_axis("keep", {"any", "first", "last", "none"})
+  .add_int64_axis("cardinality",
+                  {100, 1'000, 10'000, 100'000, 1'000'000, 10'000'000, 100'000'000, 1'000'000'000})
+  .add_int64_axis("NumRows",
+                  {100, 1'000, 10'000, 100'000, 1'000'000, 10'000'000, 100'000'000, 1'000'000'000});
```
Comment (bdice) [inline cpp/benchmarks/stream_compaction/stable_distinct.cpp:72]:
```suggestion
  .add_int64_axis("cardinality",
                  {100, 100'000, 10'000'000, 1'000'000'000})
  .add_int64_axis("NumRows",
                  {100, 100'000, 10'000'000, 1'000'000'000});
```

**Review by bdice — COMMENTED** [review]
@srinivasyadav18 Can you approve?

**Review by srinivasyadav18 — APPROVED** [review]
CI fail is still failing due styling issues.  
Otherwise LGTM! Thanks!

**Comment by bdice** [comment]
Just fixed the CI style check. I'll merge this when CI passes.

**Pull request merged**